# Modulo 4 Desarrollo de modelos 

___    
En Este [link] (https://labs.cognitiveclass.ai/tools/jupyterlab/lab/tree/labs/DA0101EN/review-introduction.ipynb) se encuentra la relacionado con el Modulo 4 del curso ciencia de datos de edx 
___  

## Temas de este modulo son
* Adquisición de datos
* Información basica del conjunto de datos


# Adquisición de Datos
Hay varios formatos para un conjunto de datos, .csv, .json, .xlsx, etc. El conjunto de datos se puede almacenar en diferentes lugares, en su máquina local o, en ocasiones, en línea.

En esta sección, aprenderá cómo cargar un conjunto de datos en nuestro Jupyter Notebook.
En nuestro caso, el conjunto de datos del automóvil es una fuente en línea y está en formato CSV (valores separados por comas). Usemos este conjunto de datos como ejemplo para practicar la lectura de datos.

* La fuente de los datos del curso de edx esta [aqui](https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data)
* el tipo de datos es csv

Pandas **Library** es una herramienta útil que nos permite leer varios conjuntos de datos en un marco de datos; 

In [1]:
#Importando la libreria de pandas
import pandas as pd

# Leer Datos
Utilizamos la función `pandas.read_csv ()` para leer el archivo csv. En el paréntesis, colocamos la ruta del archivo entre comillas, para que pandas lea el archivo en un marco de datos desde esa dirección. La ruta del archivo puede ser una URL o su dirección de archivo local.
Debido a que los datos no incluyen encabezados, podemos agregar un argumento `encabezados = Ninguno` dentro del método `read_csv ()`, para que pandas no establezcan automáticamente la primera fila como encabezado.
También puede asignar el conjunto de datos a cualquier variable que cree.

In [3]:
ruta = "imports-85.csv"

#df es traducico como dataframe (marco de datos en español) y es una variable que lee el archivo 
#que necesitamos para trabajar
#fheader=None es para que no le asigne encabezados a los datos (Titulos de la primera fila)
df = pd.read_csv(ruta,header=None)

Después de leer el conjunto de datos, podemos usar el método `dataframe.head (n)` para verificar las n filas superiores del marco de datos (las primeras n filas del archivo); donde n es un entero. Al contrario de `dataframe.head (n)`, `dataframe.tail (n)` le mostrará las n filas inferiores del marco de datos(las ultimas n filas del archivo).

In [4]:
print ("las primeras 5 filas del marco de datos son: ")
df.head(5)

las primeras 5 filas del marco de datos son: 


0    1            2    3    4     5            6    7      8     9   ...  \
0   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
1   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
2   1    ?  alfa-romero  gas  std   two    hatchback  rwd  front  94.5  ...   
3   2  164         audi  gas  std  four        sedan  fwd  front  99.8  ...   
4   2  164         audi  gas  std  four        sedan  4wd  front  99.4  ...   

    16    17    18    19    20   21    22  23  24     25  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
4  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  

[5 rows x 26 columns]

In [5]:
print ("las ultimas 10 filas del marco de datos son: ")
df.tail(10)

las ultimas 10 filas del marco de datos son: 


0    1      2       3      4     5      6    7      8      9   ...   16  \
195  -1   74  volvo     gas    std  four  wagon  rwd  front  104.3  ...  141   
196  -2  103  volvo     gas    std  four  sedan  rwd  front  104.3  ...  141   
197  -1   74  volvo     gas    std  four  wagon  rwd  front  104.3  ...  141   
198  -2  103  volvo     gas  turbo  four  sedan  rwd  front  104.3  ...  130   
199  -1   74  volvo     gas  turbo  four  wagon  rwd  front  104.3  ...  130   
200  -1   95  volvo     gas    std  four  sedan  rwd  front  109.1  ...  141   
201  -1   95  volvo     gas  turbo  four  sedan  rwd  front  109.1  ...  141   
202  -1   95  volvo     gas    std  four  sedan  rwd  front  109.1  ...  173   
203  -1   95  volvo  diesel  turbo  four  sedan  rwd  front  109.1  ...  145   
204  -1   95  volvo     gas  turbo  four  sedan  rwd  front  109.1  ...  141   

       17    18    19    20   21    22  23  24     25  
195  mpfi  3.78  3.15   9.5  114  5400  23  28  13415  
196  mpfi  3.78  3.15   9.5  114  5400  24  28  15985  
197  mpfi  3.78  3.15   9.5  114  5400  24  28  16515  
198  mpfi  3.62  3.15   7.5  162  5100  17  22  18420  
199  mpfi  3.62  3.15   7.5  162  5100  17  22  18950  
200  mpfi  3.78  3.15   9.5  114  5400  23  28  16845  
201  mpfi  3.78  3.15   8.7  160  5300  19  25  19045  
202  mpfi  3.58  2.87   8.8  134  5500  18  23  21485  
203   idi  3.01  3.40  23.0  106  4800  26  27  22470  
204  mpfi  3.78  3.15   9.5  114  5400  19  25  22625  

[10 rows x 26 columns]

In [9]:
# se crea una lista con los ecabezados, esto por que los datos no tienen (titulos en la primera fila)
#se crean para mayor entendimiento del marco de los datos
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]
print("headers\n", headers)

headers
 ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']


Echa un vistazo a nuestro conjunto de datos; pandas establece automáticamente el encabezado por un número entero desde 0.

Para describir mejor nuestros datos, podemos introducir un encabezado, esta información está disponible en: https://archive.ics.uci.edu/ml/datasets/Automobile

Por lo tanto, tenemos que agregar encabezados manualmente.

En primer lugar, creamos una lista de "encabezados" que incluyen todos los nombres de columna en orden. Luego, usamos `dataframe.columns = headers` para reemplazar los encabezados por la lista que creamos.

***Es muy importante que  los header que se creen par pasar como parametro, sean equivalentes al numero de columnas que tiene el marco de datos (df)***

Reemplazamos encabezados y volvemos a verificar nuestro marco de datos

In [10]:
df.columns = headers
df.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   
5          2                 ?         audi       gas        std          two   
6          1               158         audi       gas        std         four   
7          1                 ?         audi       gas        std         four   
8          1               158         audi       gas      turbo         four   
9          0                 ?         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   
5        sedan          fwd           front        99.8  ...          136   
6        sedan          fwd           front       105.8  ...          136   
7        wagon          fwd           front       105.8  ...          136   
8        sedan          fwd           front       105.8  ...          131   
9    hatchback          4wd           front        99.5  ...          131   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   
5         mpfi  3.19    3.40               8.5        110      5500       19   
6         mpfi  3.19    3.40               8.5        110      5500       19   
7         mpfi  3.19    3.40               8.5        110      5500       19   
8         mpfi  3.13    3.40               8.3        140      5500       17   
9         mpfi  3.13    3.40               7.0        160      5500       16   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  
5          25  15250  
6          25  17710  
7          25  18920  
8          20  23875  
9          22      ?  

[10 rows x 26 columns]

podemos "eliminar" los valores faltantes a lo largo de la columna "precio" de la siguiente manera.   
[aqui](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html?highlight=dropna#pandas.DataFrame.dropna) hay una referencia a la documentación de pandas

In [12]:
#elimina las filas que coincidan con los  campos vacios de la columna precio
df.dropna(subset = ['price'], axis = 0)


symboling normalized-losses         make fuel-type aspiration  \
0            3                 ?  alfa-romero       gas        std   
1            3                 ?  alfa-romero       gas        std   
2            1                 ?  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1            two  convertible          rwd           front        88.6  ...   
2            two    hatchback          rwd           front        94.5  ...   
3           four        sedan          fwd           front        99.8  ...   
4           four        sedan          4wd           front        99.4  ...   
..           ...          ...          ...             ...         ...  ...   
200         four        sedan          rwd           front       109.1  ...   
201         four        sedan          rwd           front       109.1  ...   
202         four        sedan          rwd           front       109.1  ...   
203         four        sedan          rwd           front       109.1  ...   
204         four        sedan          rwd           front       109.1  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0            130         mpfi  3.47    2.68               9.0        111   
1            130         mpfi  3.47    2.68               9.0        111   
2            152         mpfi  2.68    3.47               9.0        154   
3            109         mpfi  3.19    3.40              10.0        102   
4            136         mpfi  3.19    3.40               8.0        115   
..           ...          ...   ...     ...               ...        ...   
200          141         mpfi  3.78    3.15               9.5        114   
201          141         mpfi  3.78    3.15               8.7        160   
202          173         mpfi  3.58    2.87               8.8        134   
203          145          idi  3.01    3.40              23.0        106   
204          141         mpfi  3.78    3.15               9.5        114   

     peak-rpm city-mpg highway-mpg  price  
0        5000       21          27  13495  
1        5000       21          27  16500  
2        5000       19          26  16500  
3        5500       24          30  13950  
4        5500       18          22  17450  
..        ...      ...         ...    ...  
200      5400       23          28  16845  
201      5300       19          25  19045  
202      5500       18          23  21485  
203      4800       26          27  22470  
204      5400       19          25  22625  

[205 rows x 26 columns]

In [15]:
#asi obtenemos en nombre de las columnas del marco de datos
df.columns

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')

## Guardar el Conjunto de Datos
En consecuencia, Pandas nos permite guardar el conjunto de datos en `csv` utilizando el método `dataframe.to_csv ()`, puede agregar la ruta del archivo y el nombre junto con comillas entre paréntesis.   
Por ejemplo, si guarda el marco de datos df como Automobile.csv en su máquina local, puede usar la sintaxis a continuación:   
`df.to_csv("automobile.csv", index=False)`.   

También podemos leer y guardar otros formatos de archivo, podemos usar funciones similares a `pd.read_csv ()` y `df.to_csv ()` para otros formatos de datos, las funciones se enumeran en la siguiente tabla:   

| Data Formate  | Read           | Save             |
| ------------- |:--------------:| ----------------:|
| csv           | `pd.read_csv()`  |`df.to_csv()`     |
| json          | `pd.read_json()` |`df.to_json()`    |
| excel         | `pd.read_excel()`|`df.to_excel()`   |
| hdf           | `pd.read_hdf()`  |`df.to_hdf()`     |
| sql           | `pd.read_sql()`  |`df.to_sql()`     |
| ...           |   ...          |       ...        |


# Información Básica del conjunto de datos
Después de leer los datos en el marco de datos de Pandas, es hora de que exploremos el conjunto de datos. Hay varias formas de obtener información esencial de los datos para ayudarnos a comprender mejor nuestro conjunto de datos.

## Tipo de Datos
Los datos tienen una variedad de tipos. Los principales tipos almacenados en los marcos de datos de Pandas son `object`, `float`, `int`, `bool` y datetime64. Para conocer mejor cada atributo, siempre es bueno para nosotros conocer el tipo de datos de cada columna.     
En pandas:

In [16]:
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

Como resultado del comando ejecutado Anteriormente, vemos lo diferentes tipos de datos para cada columna.   
Estos tipos de datos se pueden cambiar; Aprenderemos cómo lograr esto en un módulo posterior.   
## Describir
Si nos gustaría obtener un resumen estadístico de cada columna, como conteo, valor medio de columna, desviación estándar de columna, etc. Utilizamos el método de describe(): `dataframe.describe()`


Este metodo mostrará varias estadisticas, *excluyendo* los valores NaN

In [17]:
df.describe()

symboling  wheel-base      length       width      height  \
count  205.000000  205.000000  205.000000  205.000000  205.000000   
mean     0.834146   98.756585  174.049268   65.907805   53.724878   
std      1.245307    6.021776   12.337289    2.145204    2.443522   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.300000   64.100000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.100000   66.900000   55.500000   
max      3.000000  120.900000  208.100000   72.300000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   205.000000   205.000000         205.000000  205.000000   205.000000  
mean   2555.565854   126.907317          10.142537   25.219512    30.751220  
std     520.680204    41.642693           3.972040    6.542142     6.886443  
min    1488.000000    61.000000           7.000000   13.000000    16.000000  
25%    2145.000000    97.000000           8.600000   19.000000    25.000000  
50%    2414.000000   120.000000           9.000000   24.000000    30.000000  
75%    2935.000000   141.000000           9.400000   30.000000    34.000000  
max    4066.000000   326.000000          23.000000   49.000000    54.000000

Esto muestra el resumen estadístico de todas las columnas de tipo numérico (int, float).
Por ejemplo, el atributo "simbolización" tiene 205 recuentos, el valor medio de esta columna es 0.83, la desviación estándar es 1.25, el valor mínimo es -2, el percentil 25 es 0, el percentil 50 es 1, el percentil 75 es 2 y El valor máximo es 3.
Sin embargo, ¿qué pasaría si también quisiéramos verificar todas las columnas, incluidas las que son de tipo objeto?

Para incluir todas las columnas en el metodo `describe()` se debe utilizar:

In [25]:
df.describe(include = "all")

symboling normalized-losses    make fuel-type aspiration  \
count   205.000000               205     205       205        205   
unique         NaN                52      22         2          2   
top            NaN                 ?  toyota       gas        std   
freq           NaN                41      32       185        168   
mean      0.834146               NaN     NaN       NaN        NaN   
std       1.245307               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           205        205          205             205  205.000000  ...   
unique            3          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            114         96          120             202         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.756585  ...   
std             NaN        NaN          NaN             NaN    6.021776  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    205.000000          205   205     205        205.000000        205   
unique          NaN            8    39      37               NaN         60   
top             NaN         mpfi  3.62    3.40               NaN         68   
freq            NaN           94    23      20               NaN         19   
mean     126.907317          NaN   NaN     NaN         10.142537        NaN   
std       41.642693          NaN   NaN     NaN          3.972040        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       97.000000          NaN   NaN     NaN          8.600000        NaN   
50%      120.000000          NaN   NaN     NaN          9.000000        NaN   
75%      141.000000          NaN   NaN     NaN          9.400000        NaN   
max      326.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg price  
count        205  205.000000  205.000000   205  
unique        24         NaN         NaN   187  
top         5500         NaN         NaN     ?  
freq          37         NaN         NaN     4  
mean         NaN   25.219512   30.751220   NaN  
std          NaN    6.542142    6.886443   NaN  
min          NaN   13.000000   16.000000   NaN  
25%          NaN   19.000000   25.000000   NaN  
50%          NaN   24.000000   30.000000   NaN  
75%          NaN   30.000000   34.000000   NaN  
max          NaN   49.000000   54.000000   NaN  

[11 rows x 26 columns]

Ahora, proporciona el resumen estadístico de todas las columnas, incluidos los atributos de tipo objeto.
Ahora podemos ver cuántos valores únicos, que es el valor superior y la frecuencia del valor superior en las columnas de tipo objeto.
Algunos valores en la tabla anterior se muestran como "NaN", esto se debe a que esos números no están disponibles con respecto a un tipo de columna en particular.

In [35]:
# Esta es la manera de seleccionear  columnas
df[["make", "symboling"]].describe()

symboling
count  205.000000
mean     0.834146
std      1.245307
min     -2.000000
25%      0.000000
50%      1.000000
75%      2.000000
max      3.000000

In [44]:
# Proporciona un resumen conciso del DataFrame.
df.index




RangeIndex(start=0, stop=205, step=1)

In [46]:
df.columns

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')

Otra manera de examinar el conjunto de datos es con la siguiente instrucción

In [48]:
df.info

<bound method DataFrame.info of      symboling normalized-losses         make fuel-type aspiration  \
0            3                 ?  alfa-romero       gas        std   
1            3                 ?  alfa-romero       gas        std   
2            1                 ?  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd      

In [ ]:
Aquí podemos ver la información de nuestro marco de datos, con las 30 filas superiores y las 30 filas inferiores. Y también nos muestra que el marco de datos completo tiene 205 filas y 26 columnas en total.